src/main/java/com/palantir/gradle/shadowjar/ComposableManifestAppenderTransformer.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.shadowjar;
+
+import com.github.jengelman.gradle.plugins.shadow.transformers.ResourceTransformer;
+import com.github.jengelman.gradle.plugins.shadow.transformers.TransformerContext;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+import java.util.List;
+import java.util.jar.JarFile;
+import org.apache.tools.zip.ZipEntry;
+import org.apache.tools.zip.ZipOutputStream;
+import org.gradle.api.file.FileTreeElement;
+import org.gradle.api.tasks.Input;
+
+// Originally taken from https://github.com/GradleUp/shadow/blob/9.2.2/src/main/kotlin/com/github/jengelman/gradle/
+// plugins/shadow/transformers/ManifestAppenderTransformer.kt
+public final class ComposableManifestAppenderTransformer implements ResourceTransformer {
+    private static final byte[] EOL = "\r\n".getBytes(StandardCharsets.UTF_8);
+    private static final byte[] SEPARATOR = ": ".getBytes(StandardCharsets.UTF_8);
+
+    public static final long CONSTANT_TIME_FOR_ZIP_ENTRIES =
+            new GregorianCalendar(1980, Calendar.FEBRUARY, 1, 0, 0, 0).getTimeInMillis();
+
+    private byte[] manifestContents = new byte[0];
+    private final List<Attribute> attributes = new ArrayList<>();
+
+    @Input
+    public List<Attribute> getAttributes() {
+        return attributes;
+    }
+
+    public ComposableManifestAppenderTransformer append(String name, Comparable<?> value) {
+        attributes.add(new Attribute(name, value));
+        return this;
+    }
+
+    @Override
+    public boolean canTransformResource(FileTreeElement element) {
+        return JarFile.MANIFEST_NAME.equalsIgnoreCase(element.getRelativePath().getPathString());
+    }
+
+    @Override
+    public void transform(TransformerContext context) throws IOException {
+        if (manifestContents.length == 0) {
+            try (InputStream is = context.getInputStream()) {
+                manifestContents = is.readAllBytes();
+            }
+        }
+    }
+
+    @Override
+    public boolean hasTransformedResource() {
+        return !attributes.isEmpty();
+    }
+
+    @Override
+    public void modifyOutputStream(ZipOutputStream os, boolean preserveFileTimestamps) throws IOException {
```

Comment:
same here - catching the exception and throw an UnchedIOException with extra info about the step where this fais.